Module overview 3 (#46)
apostrophecms/a3-docs: docs/reference/module-api/module-overview.md

```

### `helpers(self)`

> The `helpers` section returns an object of functions that add template utility methods. The individual helper methods may take any arguments that you plan to pass them in templates.

Reviewer: "The `helpers` section" could probably be dropped. We're already there.

        const price = product.price;
        return `$${floatPrice.toFixed(2)}`;
      }
    };
  }
};
```

Using in a template:

```django
{# modules/product-page/views/show.html #}
{{ apos.product.discountPrice(data.piece) }}
```

#### `extendHelpers(self)`

Add to the functionality of a template helper inherited from the base class. This must return an object of functions, similar to [`helpers`](#helpers-self).

Extended helpers should take a `_super` argument, followed by the normal arguments of the helper being extended. If the original helper took only a `price` argument, the extending function should take the arguments `_super, price`.

To maintain the same application, they should return the same type of response as the original helper. If the original returned a string, the extension should return a string.

```javascript
// modules/featured-product/index.js
module.exports = {
  extend: 'product',
  // ...
  extendHelpers(self) {
    return {
      formatPrice(_super, product) {
        const price = _super(product);
        // 👇 Adds some extra flash to the featured product prices.
        return `${price} 🤑`;
      }
    };
  }
};
```

### `restApiRoutes(self)`

> Apostrophe includes a full REST API for [pieces](/reference/api/pieces.md) and [pages](/reference/api/pages.md). Those route handlers can be totally overridden by adding identically named functions to object returned by the `restApiRoutes` function. If you simply wish to add to the existing behavior of the REST API routes, see [`extendRestApiRoutes`](#extendrestapiroutes-self).

Reviewer: "**Those route handlers can be totally overridden** by adding identically named functions **to object** returned by the `restApiRoutes` function. If you **simply wish to add to** the existing behavior of the REST API routes, see [`extendRestApiRoutes`](#extendrestapiroutes-self)."

"Those route handlers can be totally overridden" -> "You can completely override these route handlers" (completely optional?)

"to object" -> "to objects"

"If you **simply wish to add to**" -> "To add to and extend existing behavior"

Reviewer: I love wishes tho

Author: I'm redoing that whole section based on Tom's notes.

> Apostrophe includes a full REST API for [pieces](/reference/api/pieces.md) and [pages](/reference/api/pages.md). Those route handlers can be totally overridden by adding identically named functions to object returned by the `restApiRoutes` function. If you simply wish to add to the existing behavior of the REST API routes, see [`extendRestApiRoutes`](#extendrestapiroutes-self).

Reviewer 2: You imply it, but this really needs to be emphasized:

All piece type modules, as well as the `@apostrophecms/page` module, already have a complete REST API and you should not use `restApiRoutes` with these. The most likely use of `restApiRoutes` in a project would be to provide a custom REST API to a database or service not already built into Apostrophe. `restApiRoutes` is also not for custom routes that don't map to one of the standard REST URLs. If you need to add a custom route over and above the REST API of pieces or pages, you should do that with `apiRoutes`.

Author: > The most likely use of restApiRoutes in a project of your own would be to provide a custom REST API to a database or service not already built into Apostrophe.

@boutell Could you suggest a more specific use case? I want to make sure I'm communicating what you're saying correctly. Like I create a `rando` module that is not a piece, then I add a REST API that actually hits a third-party database?

Author: Maybe `apiRoutes` should appear first in the page.

REST API functions take the route request as an argument (`req`, below);
Valid names for functions returned by `restApiRoutes` include:
- `getAll`
- `getOne`
- `post`
- `patch`
- `put`
- `delete`

```javascript
// modules/product/index.js
module.exports = {
  // ...
  restApiRoutes(self) {
    return {
      // GET /api/v1/product
      async getAll(req) {
        const results = [];
        // ... populate `results` with product data.

        return {
          results
        };
      }
    };
  }
};
```

#### `extendRestApiRoutes(self)`

> Extend the behavior of existing REST API routes in the `extendRestApiRoutes` section. This function must return an object of functions. See [`restApiRoutes`](#restapiroutes-self) for the valid function names.

Reviewer: "in the `extendRestApiRoutes` section" -> remove

Author: These should say "function" not "section."

Author: Or nothing.

          // Adds a `resultLength` property on the response object.
          response.resultLength = response.results.length;
        }

        return response;
      }
    };
  }
};
```

### `apiRoutes(self)`

Add custom API routes. The `apiRoutes` function takes takes the module as an argument and must return an object with properties for the relevant [HTTP request method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods), including `get`, `post`, `patch`, and `delete`. Each of those properties should be set to an object of functions.

```javascript
// modules/product/index.js
module.exports = {
  // ...
  apiRoutes(self) {
    return {
      get: {
        // GET /api/v1/product/cheapest
        async cheapest(req) {
          const product = await self.find(req).sort({
            price: 1
          }).toObject();
          if (!product) {
            // Browser receives a 404 error
            throw self.apos.error('notfound', 'No products were found.');
          }
          // Response is a JSON object with a `product` property
          return {
            product
          };
        }
      }
    };
  }
};
```

#### Naming routes

> If route properties do *not* begin with a forward slash, they can be reached via the pattern: `/api/v1/` followed by the module name, a forward slash, and the API route handler's property name, e.g., `/api/v1/product/cheapest` for the `cheapest` route on the `product` module.

Reviewer: `cheapest` could probably be replaced by something that translates a bit better, since its pretty much slang. Using product....my first thought is `brand/s` or a category like `denim-jacket/s`. Denim and jacket are both fairly universal, while being very American!

Bonus points if we find a way to use infant bodysuit

Author: 😕 "cheapest" simply means "most cheap." Least expensive.

You can also use a completely custom URL path by starting the name of the route with a forward slash (`/`). If you do so, nothing will be prefixed to it and it will not be converted to kebab case.

```javascript
// modules/product/index.js
module.exports = {
  // ...
  apiRoutes(self) {
    return {
      get: {
        // GET /api/v1/product/cheapest-one
        async cheapestOne(req) {
          // ...
        },
        // GET /my-api/cheapest
        '/my-api/cheapest': async function (req) {
          // ...
        }
      }
    };
  }
};
```

#### Returning error codes

You can return the `self.apos.error()` method in a route function to return specific error codes. Pass in one of several strings to set a specific error response code:

| Error name | HTTP response code |
| ---- | ---- |
| `'invalid'` | 400 |
| `'forbidden'` | 403 |
| `'notfound'` | 404 |
| `'required'` | 422 |
| `'conflict'` | 409 |
| `'locked'` | 409 |
| `'unprocessable'` | 422 |
| `'unimplemented'` | 501 |

Passing a different value as the first argument to `self.apos.error()` will set the response code to 500.

> <!-- TODO: Link to the method's own documentation page when available for more. -->

Author: I'm holding off on full documentation of that here since it could be a big chunk. I'm going to link over to it once available instead. There'll be a sentence like "See more information on using the `error()` method."

#### `extendApiRoutes(self)`

> Extend the behavior of existing API routes (set in `apiRoutes`) in the `extendApiRoutes` section. This function must return an object as described in [`apiRoutes`](#apiroutes-self).

Reviewer: "in the `extendApiRoutes` section" -> dont need (imo)

Each extended API route function should accept the original function as `_super` and the `req` request object. They should return data in a similar format to the existing API route.

```javascript
// modules/featured-product/index.js
module.exports = {
  extend: 'product',
  // ...
  extendApiRoutes(self) {
    return {
      get: {
        // GET /api/v1/featured-product/cheapest
        async cheapest(_super, req) {
          const response = _super(req);

          // Update the response object...

          return response;
        }
      }
    };
  }
};
```

### `renderRoutes(self)`

> Add custom API routes to return rendered templates. The `apiRoutes` function takes takes the module as an argument and must return an object with properties for the relevant [HTTP request method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods), including `get`, `post`, `patch`, and `delete`. Each of those properties should be set to an object of functions.

Reviewer: "the module as an argument" -> is this the module name?

Author: No, it's the actual module. That's what `self` is.

The name of the route dictates the template file that will be rendered. For example the `latest` route in the `product` module's `renderRoutes` section will return the template at `/modules/product/views/latest.html`.

Information returned by the component function will be used in the associated template as `data`.

```javascript
// modules/product/index.js
module.exports = {
  // ...
  renderRoutes(self) {
    return {
      get: {
        // GET /api/v1/product/latest
        // The route rendered HTML for /modules/product/views/latest.html
        async latest(req) {
          const products = await self.find(req)
            .sort({ createdAt: -1 })
            .limit(req.query.max || 5)
            .toArray();

          return {
            products
          };
        }
      }
    };
  }
};
```

### `routes(self)`

Add standard Express routes. The `routes` function takes takes the module as an argument and must return an object with properties for the relevant [HTTP request method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods), including `get`, `post`, `patch`, and `delete`. Each of those properties should be set to an object of functions.

> Each `routes` function takes the Express arguments `req` (the [request object](https://expressjs.com/en/api.html#req)) and `res` (the [response object](https://expressjs.com/en/api.html#res)). The functions must return using the `res.redirect` or `res.send` methods.

Reviewer 2: How about: "the functions must generate a response via `res` to avoid leaking resources, typically using the `res.redirect` or `res.send` methods." There are other, less often used ways to end a response, such as calling `res.send`, so I don't want to make it sound like Apostrophe is doing anything other than letting you do as you like with Express here.

Reviewer 2: `routes` should be used only where `apiRoutes` or `restApiRoutes` or `renderRoutes` won't do, such as when you need to use `res.redirect` or pipe a stream. We should state this clearly to help guide people to the painless solution. Failing to do the right thing with `res` or handle errors etc. is a common cause of Node.js bugs, which is why apiRoutes exists.

(At some point I'd like to add a way to use `res.redirect` via a return value. Piping a stream will probably always be an "uh... just use routes()" scenario.")

Author: I've added a new :::tip block a few lines down.

See [Naming routes](#naming-routes) for more on function names and their route URLs.

```javascript
// modules/product/index.js
module.exports = {
  // ...
  renderRoutes(self) {
    return {
      get: {
        // GET /api/v1/product/redirect/:_id
        async redirect(req, res) {
          const product = await self.find(req).toObject();

          return res.redirect(product._url);
        }
      }
    };
  }
};
```

### `handlers(self)`

The `handlers` function takes the module as an argument and must return an object. The object keys should be names of existing server-side events. The value of those event keys should be an object of functions to execute when those events fire.
<!-- TODO: Link to the reference to or guide on server-side events when available. -->

> Events belonging to the same module where the handlers are defined, or from its base class, can be referenced simply by name, e.g., `beforeInsert` for any piece type. You may also add handlers in one module that respond to events in other modules. Those event names should be prefixed with the name of the module where the event fires followed by a colon, e.g., `@apostrophecms/page:beforeSend`.

Reviewer 2: Event handlers may be async functions, and they will be awaited before execution continues.

Author: I've been internally debating how much to explicitly say that things can be async. It's probably a good idea to do more.

> Events belonging to the same module where the handlers are defined, or from its base class, can be referenced simply by name, e.g., `beforeInsert` for any piece type. You may also add handlers in one module that respond to events in other modules. Those event names should be prefixed with the name of the module where the event fires followed by a colon, e.g., `@apostrophecms/page:beforeSend`.

Reviewer: "simply by name" -> "by name"

```

### `helpers(self)`

> `helpers` returns an object of functions that add template utility methods. The individual helper methods may take any arguments that you plan to pass them in templates. Helper functions must run synchronously.

Reviewer 2: How about: "the individual helper methods may accept arguments when called from your templates."

  // ...
  apiRoutes(self) {
    return {
      get: {
        // GET /api/v1/product/cheapest-one
        async cheapestOne(req) {
          // ...
        },
        // GET /my-api/cheapest
        '/my-api/cheapest': async function (req) {
          // ...
        }
      }
    };
  }
};
```

#### Returning error codes

You can `throw` the `self.apos.error()` method in any route function to return specific error codes as well as log additional information for developers. Pass in one of several strings to set a specific error response code:

| Error name | HTTP response code |
| ---- | ---- |
| `'invalid'` | 400 |
| `'forbidden'` | 403 |
| `'notfound'` | 404 |
| `'required'` | 422 |
| `'conflict'` | 409 |
| `'locked'` | 409 |
| `'unprocessable'` | 422 |
| `'unimplemented'` | 501 |

Passing a different value as the first argument to `self.apos.error()` will set the response code to 500.
<!-- TODO: Link to the method's own documentation page when available for more. -->

#### `extendApiRoutes(self)`

Extend the behavior of existing API routes (set in `apiRoutes`) in `extendApiRoutes`. This function must return an object as described in [`apiRoutes`](#apiroutes-self).

> Each extended API route function should accept the original function as `_super` and the `req` request object. They should return data in a similar format to the existing API route.

Reviewer 2: "Your extended API route function should"?

        // The route rendered HTML for /modules/product/views/latest.html
        async latest(req) {
          const products = await self.find(req)
            .sort({ createdAt: -1 })
            .limit(req.query.max || 5)
            .toArray();

          return {
            products
          };
        }
      }
    };
  }
};
```

### `routes(self)`

Add standard Express routes. The `routes` function takes takes the module as an argument and must return an object with properties for the relevant [HTTP request method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods), including `get`, `post`, `patch`, and `delete`. Each of those properties should be set to an object of functions. Route functions may be asynchronous (async).

Each `routes` function takes the Express arguments `req` (the [request object](https://expressjs.com/en/api.html#req)) and `res` (the [response object](https://expressjs.com/en/api.html#res)). The functions must generate a response via `res` to avoid leaking resources, typically using the `res.redirect` or `res.send` methods.

See [Naming routes](#naming-routes) for more on function names and their route URLs.

::: tip

> We recommend using `apiRoutes` or `restApiRoutes` whenever possible before using `routes`. They are designed to be easier to use and handle the potential pitfalls of Express routes. There are situation when writing Express routes may be necessary, such as when you need to use `res.redirect` or pipe a stream.

Reviewer: ```suggestion
We recommend using `apiRoutes` or `restApiRoutes` whenever possible before using `routes`, as they handle potential pitfalls of Express routes. There are situations where writing Express routes may be necessary, such as when you need to use `res.redirect` or pipe a stream.
```

:::

```javascript
// modules/product/index.js
module.exports = {
  // ...
  renderRoutes(self) {
    return {
      get: {
        // GET /api/v1/product/redirect/:_id
        async redirect(req, res) {
          const product = await self.find(req).toObject();

          return res.redirect(product._url);
        }
      }
    };
  }
};
```

### `handlers(self)`

The `handlers` function takes the module as an argument and must return an object. The object keys should be names of existing server-side events. The value of those event keys should be an object of functions to execute when those events fire. Event handlers may be asynchronous (async) functions.
<!-- TODO: Link to the reference to or guide on server-side events when available. -->

> Events belonging to the same module where the handlers are defined, or from its base class, can be referenced by name, e.g., `beforeInsert` for any piece type. You may also add handlers in one module that respond to events in other modules. Those event names should be prefixed with the name of the module where the event fires followed by a colon, e.g., `@apostrophecms/page:beforeSend`.

Reviewer: ```suggestion
Events belonging to the same module where the handlers are defined, or from its base class, can be referenced by name, e.g., `beforeInsert` for any piece type. You may also add `handlers` in one module that respond to events in other modules. Those event names should be prefixed with the name of the module where the event occurs, followed by a colon, e.g., `@apostrophecms/page:beforeSend`.
```

Author: "handlers" is also the name of the type of function, like we call then "Event handlers" earlier.
